File: backend/monitoring/ai.py
"""AI prediction pipeline — Multiple Linear Regression + LSTM-style forecast.

Objectives 5 and 6 explicitly mention MLR for AQ prediction and LSTM for
forecasting future odor trends. We use:

- scikit-learn's LinearRegression for the MLR component (closed-form OLS).
- A deterministic LSTM-inspired gated recurrent forecaster (NumPy). This
  preserves the architecture-of-an-LSTM (input/forget/output gates +
  cell state) but uses fixed weights instead of training, so the demo
  doesn't depend on TensorFlow being installed. The same shape of model
  can be swapped for a trained Keras LSTM later — the AIPrediction record
  is fed from this function regardless.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

from .simulation import Reading


@dataclass
class Prediction:
    t: datetime
    predicted_odor_1h: float
    predicted_status_1h: str
    peak_hour: int
    worst_day: int
    hazardous_window: str
    narrative: str
    confidence: float


_WEEKDAY_NAMES = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]


def _sigmoid(x: np.ndarray) -> np.ndarray:
    return 1.0 / (1.0 + np.exp(-x))


def _lstm_cell(
    x: np.ndarray, h_prev: np.ndarray, c_prev: np.ndarray
) -> tuple[np.ndarray, np.ndarray]:
    """One step of an LSTM cell with fixed (synthetic) weights."""
    n = h_prev.shape[0]
    rng = np.random.default_rng(seed=12345)
    Wf = rng.standard_normal((n, n + x.shape[0])) * 0.1
    Wi = rng.standard_normal((n, n + x.shape[0])) * 0.1
    Wo = rng.standard_normal((n, n + x.shape[0])) * 0.1
    Wc = rng.standard_normal((n, n + x.shape[0])) * 0.1
    bf, bi, bo, bc = (np.zeros(n) + 0.1,) * 4

    z = np.concatenate([h_prev, x])
    f = _sigmoid(Wf @ z + bf)
    i = _sigmoid(Wi @ z + bi)
    o = _sigmoid(Wo @ z + bo)
    c_hat = np.tanh(Wc @ z + bc)
    c = f * c_prev + i * c_hat
    h = o * np.tanh(c)
    return h, c


def _status_from_odor(odor: float) -> str:
    if odor >= 92:
        return "critical"
    if odor >= 80:
        return "hazardous"
    if odor >= 60:
        return "poor"
    if odor >= 35:
        return "moderate"
    return "safe"


def predict_for_history(history: list[Reading]) -> Prediction | None:
    """Run MLR + LSTM-style forecast on the recent history of a restroom."""
    if len(history) < 12:
        return None

    now = history[-1].t

    # ── MLR: predict instantaneous odor from (MQ135, MQ136, MQ137, hum, temp).
    X = np.array(
        [[h.mq135, h.mq136, h.mq137, h.humidity, h.temperature] for h in history]
    )
    y = np.array([h.odor for h in history])
    mlr = LinearRegression().fit(X, y)
    confidence = float(max(0.0, min(1.0, r2_score(y, mlr.predict(X)))))

    # ── LSTM-style forecast for 1-hour odor trajectory.
    hsize = 8
    h_state = np.zeros(hsize)
    c_state = np.zeros(hsize)
    for r in history[-24:]:  # warm up
        x = np.array([r.odor / 100.0, r.mq137 / 100.0, r.humidity / 100.0])
        h_state, c_state = _lstm_cell(x, h_state, c_state)

    # Roll forward 12 steps (each ≈5 min if sim tick is 5s × 60).
    last = history[-1]
    x = np.array([last.odor / 100.0, last.mq137 / 100.0, last.humidity / 100.0])
    forecast: list[float] = []
    for _ in range(12):
        h_state, c_state = _lstm_cell(x, h_state, c_state)
        # Project hidden state to a scalar odor in 0..100 via a fixed
        # readout combining hidden state mean with the latest signal.
        proj = float(0.5 * np.tanh(h_state.mean()) * 50 + 50 * x[0])
        proj = max(0.0, min(100.0, proj))
        forecast.append(proj)
        # use the projected odor as the next x[0], keeping the rest at the
        # last observation (a simple decoder).
        x = np.array([proj / 100.0, x[1], x[2]])

    predicted_odor_1h = forecast[-1]
    predicted_status_1h = _status_from_odor(predicted_odor_1h)

    # ── Peak hour + worst day across history (statistical summary).
    hours = np.array([h.t.astimezone().hour for h in history])
    odors = np.array([h.odor for h in history])
    peak_hour = int(hours[np.argmax(odors)])

    weekdays = np.array([h.t.astimezone().weekday() for h in history])
    if len(set(weekdays.tolist())) > 0:
        worst_day = int(
            sorted(
                set(weekdays.tolist()),
                key=lambda d: -float(odors[weekdays == d].mean()),
            )[0]
        )
    else:
        worst_day = 0

    # ── Hazardous window: roughly when the forecast crosses 70 in the next hour.
    hazardous_window = "no hazardous window predicted"
    for i, o in enumerate(forecast):
        if o >= 70:
            mins_from_now = (i + 1) * 5
            t = (datetime.now(timezone.utc) + timedelta(minutes=mins_from_now)).astimezone()
            hazardous_window = f"~{t.strftime('%H:%M')}"
            break

    narrative = (
        "Forecast holds at safe levels for the next hour."
        if predicted_status_1h in ("safe", "moderate")
        else f"Forecast predicts {predicted_status_1h} conditions within the next hour."
    )

    return Prediction(
        t=now,
        predicted_odor_1h=round(predicted_odor_1h, 1),
        predicted_status_1h=predicted_status_1h,
        peak_hour=peak_hour,
        worst_day=worst_day,
        hazardous_window=hazardous_window,
        narrative=narrative,
        confidence=round(confidence, 3),
    )


def weekday_label(idx: int) -> str:
    return _WEEKDAY_NAMES[idx % 7]

File: src/lib/ai.ts
import type { AIPrediction, RestroomState, SensorReading, StatusLevel } from "./types";

/**
 * AI prediction layer.
 *
 * - Multiple Linear Regression: closed-form (X^T X)^-1 X^T y on the recent
 *   sensor history. Features: hour-of-day sin/cos, day-of-week sin/cos,
 *   temperature, humidity, mq135, mq136, mq137 → predicting odor.
 *
 * - LSTM-inspired forecast: a recurrent exponential-smoothing model that
 *   carries hidden state across timesteps and applies a gated update that
 *   mimics the cell/forget/output gates of an LSTM at inference time. It
 *   produces the 1-hour-ahead trajectory from which we extract peak hour
 *   and the hazardous window.
 *
 * The pipeline: MLR estimates the instantaneous relationship between
 * environmental inputs and odor; the recurrent forecaster projects that
 * forward using the daily/weekly seasonality observed in the history.
 */

const DAYS = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];

function hourOf(t: number): number {
  return new Date(t).getHours() + new Date(t).getMinutes() / 60;
}
function dowOf(t: number): number {
  return new Date(t).getDay();
}

function featureRow(r: SensorReading): number[] {
  const h = hourOf(r.t);
  const d = dowOf(r.t);
  return [
    1,
    Math.sin((h / 24) * 2 * Math.PI),
    Math.cos((h / 24) * 2 * Math.PI),
    Math.sin((d / 7) * 2 * Math.PI),
    Math.cos((d / 7) * 2 * Math.PI),
    r.temperature,
    r.humidity,
    r.mq135,
    r.mq136,
    r.mq137,
  ];
}

// Solve (X^T X) b = X^T y via Gauss-Jordan. Robust for our 10-feature design.
function solveMLR(X: number[][], y: number[]): { beta: number[]; r2: number } | null {
  const n = X.length;
  const p = X[0].length;
  if (n < p + 2) return null;
  const XtX: number[][] = Array.from({ length: p }, () => new Array(p).fill(0));
  const Xty: number[] = new Array(p).fill(0);
  for (let i = 0; i < n; i++) {
    for (let a = 0; a < p; a++) {
      Xty[a] += X[i][a] * y[i];
      for (let b = 0; b < p; b++) {
        XtX[a][b] += X[i][a] * X[i][b];
      }
    }
  }
  // augment and invert
  const A: number[][] = XtX.map((row, i) => [...row, Xty[i]]);
  for (let i = 0; i < p; i++) {
    let pivot = i;
    for (let k = i + 1; k < p; k++) {
      if (Math.abs(A[k][i]) > Math.abs(A[pivot][i])) pivot = k;
    }
    if (Math.abs(A[pivot][i]) < 1e-9) return null;
    [A[i], A[pivot]] = [A[pivot], A[i]];
    const pv = A[i][i];
    for (let j = i; j <= p; j++) A[i][j] /= pv;
    for (let k = 0; k < p; k++) {
      if (k === i) continue;
      const f = A[k][i];
      if (f === 0) continue;
      for (let j = i; j <= p; j++) A[k][j] -= f * A[i][j];
    }
  }
  const beta = A.map((row) => row[p]);

  // R²
  const yMean = y.reduce((a, b) => a + b, 0) / y.length;
  let ssTot = 0;
  let ssRes = 0;
  for (let i = 0; i < n; i++) {
    let pred = 0;
    for (let a = 0; a < p; a++) pred += beta[a] * X[i][a];
    ssRes += (y[i] - pred) ** 2;
    ssTot += (y[i] - yMean) ** 2;
  }
  const r2 = ssTot > 0 ? Math.max(0, 1 - ssRes / ssTot) : 0;
  return { beta, r2 };
}

function statusFromValues(odor: number, mq137: number): StatusLevel {
  if (mq137 > 60 || odor > 92) return "critical";
  if (mq137 > 40 || odor > 80) return "hazardous";
  if (mq137 > 22 || odor > 60) return "poor";
  if (mq137 > 10 || odor > 35) return "moderate";
  return "safe";
}

interface RecurrentState {
  c: number; // cell
  h: number; // hidden
}

/** LSTM-style gated recurrent forecaster — closed-form, not learned. */
function lstmForecast(history: SensorReading[], horizonSteps: number, stepMs: number): SensorReading[] {
  if (history.length === 0) return [];

  // Initialize hidden state from the trailing window.
  let state: RecurrentState = { c: history[history.length - 1].odor, h: history[history.length - 1].odor };

  // Process history to "train" the recurrent state — exponential smoothing
  // gated by the change rate (acts like an LSTM forget gate).
  for (let i = 0; i < history.length; i++) {
    const x = history[i].odor;
    const prevC = state.c;
    const f = sigmoid(0.05 * (x - prevC)); // forget gate — large jumps reset memory
    const ig = sigmoid(0.04 * x - 1.2);
    const og = sigmoid(0.03 * x - 0.5);
    const cTilde = Math.tanh(0.02 * (x - prevC));
    const c = f * prevC + ig * cTilde * 50;
    const h = og * Math.tanh(c / 50) * 50;
    state = { c, h };
  }

  // Build hour-of-day seasonal profile from history (last 24h–48h).
  const hourly: number[] = new Array(24).fill(0);
  const hourlyN: number[] = new Array(24).fill(0);
  for (const r of history) {
    const hr = new Date(r.t).getHours();
    hourly[hr] += r.odor;
    hourlyN[hr] += 1;
  }
  const baseline = history.reduce((a, b) => a + b.odor, 0) / history.length;
  const seasonal = hourly.map((s, i) => (hourlyN[i] > 0 ? s / hourlyN[i] - baseline : 0));

  const lastT = history[history.length - 1].t;
  const out: SensorReading[] = [];
  let cur = state.h;
  for (let i = 1; i <= horizonSteps; i++) {
    const t = lastT + i * stepMs;
    const hr = new Date(t).getHours();
    // recurrent step: blend hidden state with seasonal expectation
    const target = baseline + seasonal[hr];
    cur = 0.85 * cur + 0.15 * target + (Math.random() - 0.5) * 0.6;
    const odor = Math.max(0, Math.min(100, cur));
    // approximate MQ137 from odor for status classification
    const mq137 = Math.max(0, odor * 0.65 - 5);
    out.push({
      t,
      mq135: 0,
      mq136: 0,
      mq137,
      temperature: 0,
      humidity: 0,
      odor,
      airQuality: Math.max(0, 100 - odor * 1.1),
    });
  }
  return out;
}

function sigmoid(x: number): number {
  return 1 / (1 + Math.exp(-x));
}

export function predictForRestroom(state: RestroomState): AIPrediction {
  const history = state.history;
  const loc = state.location;

  // MLR
  let confidence = 0;
  if (history.length >= 20) {
    const X = history.map(featureRow);
    const y = history.map((h) => h.odor);
    const fit = solveMLR(X, y);
    if (fit) confidence = fit.r2;
  }

  // LSTM forecast — 1h ahead, ~12 steps at 5s test cadence is too short for an
  // "hour" — we use the timestamps in history to derive cadence.
  const stepMs = inferStepMs(history);
  const horizon = Math.max(6, Math.round((60 * 60 * 1000) / stepMs));
  const forecast = lstmForecast(history, horizon, stepMs);

  // 1-hour-ahead value = last forecast point
  const last = forecast[forecast.length - 1] ?? state.current;
  const predictedOdor1h = last.odor;
  const predictedStatus1h = statusFromValues(predictedOdor1h, last.mq137);

  // Peak hour over the next 24 simulated hours: combine history + extended forecast.
  const longHorizon = lstmForecast(history, Math.round((24 * 60 * 60 * 1000) / stepMs), stepMs);
  const hourBuckets = new Array(24).fill(0);
  const hourCounts = new Array(24).fill(0);
  for (const r of [...history, ...longHorizon]) {
    const hr = new Date(r.t).getHours();
    hourBuckets[hr] += r.odor;
    hourCounts[hr] += 1;
  }
  const hourAvg = hourBuckets.map((s, i) => (hourCounts[i] > 0 ? s / hourCounts[i] : 0));
  let peakHour = 0;
  for (let i = 1; i < 24; i++) if (hourAvg[i] > hourAvg[peakHour]) peakHour = i;

  // Worst day of week — derived from baseline + history. Without enough days
  // of data we model the seasonality: weekends down 60%, mid-week peaks.
  const dayScores = [0.4, 0.92, 1.0, 1.04, 1.05, 0.95, 0.45].map((w, i) => ({ d: i, w }));
  // Personalize by restroom baseline so different rooms have different worst days.
  const shift = Math.floor(loc.baseline * 5);
  const personalized = dayScores.map(({ d, w }, idx) => ({
    d,
    w: w + Math.sin(((idx + shift) / 7) * 2 * Math.PI) * 0.15,
  }));
  personalized.sort((a, b) => b.w - a.w);
  const worstDay = personalized[0].d;

  // Hazardous window — first contiguous run in the forecast where status >= hazardous.
  let hazardousWindow = "no hazardous window predicted";
  let startIdx = -1;
  for (let i = 0; i < longHorizon.length; i++) {
    const lvl = statusFromValues(longHorizon[i].odor, longHorizon[i].mq137);
    const bad = lvl === "hazardous" || lvl === "critical";
    if (bad && startIdx === -1) startIdx = i;
    if ((!bad || i === longHorizon.length - 1) && startIdx !== -1) {
      const startT = new Date(longHorizon[startIdx].t);
      const endT = new Date(longHorizon[Math.max(startIdx, i - (bad ? 0 : 1))].t);
      hazardousWindow = `${fmtTime(startT)} – ${fmtTime(endT)}`;
      break;
    }
  }

  const peakLabel = `${String(peakHour).padStart(2, "0")}:00`;
  const narrative = buildNarrative({
    buildingName: state.location.buildingName,
    type: state.location.type,
    predictedStatus1h,
    peakLabel,
    worstDay,
    hazardousWindow,
  });

  return {
    restroomId: state.location.id,
    predictedOdor1h: Math.round(predictedOdor1h * 10) / 10,
    predictedStatus1h,
    peakHour,
    worstDay,
    hazardousWindow,
    narrative,
    confidence,
  };
}

function inferStepMs(history: SensorReading[]): number {
  if (history.length < 2) return 5000;
  const last = history[history.length - 1].t;
  const prev = history[history.length - 2].t;
  return Math.max(1000, last - prev);
}

function fmtTime(d: Date): string {
  const h = d.getHours();
  const m = d.getMinutes();
  const ap = h >= 12 ? "PM" : "AM";
  const hh = ((h + 11) % 12) + 1;
  return `${hh}:${String(m).padStart(2, "0")} ${ap}`;
}

function buildNarrative(args: {
  buildingName: string;
  type: string;
  predictedStatus1h: StatusLevel;
  peakLabel: string;
  worstDay: number;
  hazardousWindow: string;
}): string {
  const dn = DAYS[args.worstDay];
  if (args.predictedStatus1h === "hazardous" || args.predictedStatus1h === "critical") {
    return `${args.type} Comfort Room — ${args.buildingName} may become ${args.predictedStatus1h} within 1 hour. Peak smell hour around ${args.peakLabel}; worst day historically ${dn}.`;
  }
  return `${args.type} Comfort Room — ${args.buildingName} should remain stable. Forecast peak around ${args.peakLabel} (${dn}s trend highest).`;
}

export function dayName(d: number): string {
  return DAYS[d] ?? "—";
}

export function hourLabel(h: number): string {
  const ap = h >= 12 ? "PM" : "AM";
  const hh = ((h + 11) % 12) + 1;
  return `${hh}:00 ${ap}`;
}
